src/model/markercollection.js

```diff
@@ -0,0 +1,287 @@
+		this.fire( 'add', marker );
+
+		return marker;
+	}
+
+	/**
+	 * Removes given {@link ~Marker marker} or a marker with given name from the `MarkerCollection`.
+	 *
+	 * @param {String} markerOrName Marker or name of a marker to remove.
+	 * @returns {Boolean} `true` if marker was found and removed, `false` otherwise.
+	 */
+	remove( markerOrName ) {
+		const markerName = markerOrName instanceof Marker ? markerOrName.name : markerOrName;
+		const marker = this._markers.get( markerName );
+
+		if ( marker ) {
+			this._markers.delete( markerName );
+			this.fire( 'remove', marker );
+
+			this._destroyMarker( marker );
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Destroys markers collection.
+	 */
+	destroy() {
+		for ( let marker of this._markers.values() ) {
+			this._destroyMarker( marker );
+		}
+
+		this._markers = null;
+
+		this.stopListening();
+	}
+
+	/**
+	 * Destroys marker.
+	 *
+	 * @private
+	 * @param {~Marker} marker Marker to destroy.
+	 */
+	_destroyMarker( marker ) {
+		marker.stopListening();
+		marker._liveRange.detach();
+		marker._liveRange = null;
+	}
+
+	/**
+	 * Fired whenever marker is added to `MarkerCollection`.
+	 *
+	 * @event add
+	 * @param {~Marker} The added marker.
+	 */
+
+	/**
+	 * Fired whenever marker is removed from `MarkerCollection`.
+	 *
+	 * @event remove
+	 * @param {~Marker} marker The removed marker.
+	 */
+}
+
+mix( MarkerCollection, EmitterMixin );
+
+/**
+ * `Marker` is a continuous parts of model (like a range), is named and represent some kind of information about marked
+ * part of model document. In contrary to {@link module:engine/model/node~Node nodes}, which are bits of data, markers are
+ * kind of bookmarks, marking part of a model document between marker start and marker end.
+ *
+ * Since markers are based on {@link module:engine/model/liverange~LiveRange live ranges}, for efficiency reasons, it's
+ * best to create and keep at least markers as possible.
+ *
+ * Markers can be converted to view by adding appropriate converters for
+ * {@link module:engine/conversion/modelconversiondispatcher~ModelConversionDispatcher#event:addMarker addMarker} and
+ * {@link module:engine/conversion/modelconversiondispatcher~ModelConversionDispatcher#event:removeMarker removeMarker}
+ * events, or by building converters for {@link module:engine/conversion/modelconversiondispatcher~ModelConversionDispatcher}
+ * using {@link module:engine/conversion/buildmodelconverter~buildModelConverter model converter builder}.
+ *
+ * Markers are similar to adding and converting attributes on nodes. The difference is that attribute is connected with
+ * a given node (e.g. a character is bold no matter if it gets moved or content around it changes). Markers on the
+ * other hand are continuous ranges (e.g. a character inside of marker doesn't have any special property itself, when it
+ * is moved outside of marker it stops being "special" and the marker is shrunk).
```

Comment:
Also, to find attribute you need to scan the whole document, while with a marker you always know which range is marked.